### PR TITLE
feat: add config and option flow, multiple hubs

### PR DESCRIPTION
**Pull request by argoyle**
- Add config flow
- Add option flow to make update interval configurable
- Add unique-ids
- Add possibility to unload/reload configuration
- Add support for multiple configurations (if someone has more than one EnergyHub)
- Add support for restoring entity state
- Add device info to entities (i.e. grouping them by EnergyHub, SSO, ESO and ESM)
- Average values of data received during update interval where possible
- Prepare for HACS-installation by changing the directory structure

### CODE REVIEW

**Comment by henricm** [comment]
This is awesome. Looks like giant leap for this component and I will also learn a lot from these changes. Thanks a lot!

Would it be possible to preserve the history for the files that have been moved into `custom_components` (`git mv` instead of `rm` and `add`)? That will help a lot tracking the history and changes. Thanks!

I hope to be able to test this during the coming week.

**Comment by argoyle** [comment]
Keep in mind that I'm no Python-developer, just dipping my toes in right now. 😊 So if there is anything Python-wise that can be done in a better way, please add comments. I'll see what I can do about the add/rm things.

**Comment by argoyle** [comment]
We should probably add some note for people migrating from the previous version since I think the best way is to remove the integration, restart HA and then update and add the integration again since entities might get _2 id's otherwise. I think most entities has the same id's still.

**Comment by argoyle** [comment]
There, re-did the restructure as a separate commit. Will update readme a bit more tomorrow and might add some tests. Doing tests for config and option flow now. Might put that in a separate PR afterwards though.

**Comment by argoyle** [comment]
Ok, updated the readme about upgrading as well now.

**Comment by argoyle** [comment]
I seem to get a ghost-device after a while which I don't understand. Please keep an eye out for what could be the cause of that. 😊 
image

**Comment by argoyle** [comment]
Added some logging and it seems like a message without id is sent which then cause my ghost to show up of course:
```
Received ESO-event without an id: '{"soc": {"val": 0.0}, "temp": {"val": 0.0}, "wbatcons": {"val": 0}, "ubat": {"val": 0.0}, "ibat": {"val": 0.0}, "relaystatus": {"val": "0"}, "faultcode": {"val": "0"}, "ts": {"val": "2021-03-03T01:03:55UTC"}, "id": {"val": ""}, "wbatprod": {"val": 0}}'
```

I'll add an ignore for those I think.

File: README.md [hunk 1]
```diff
@@ -7,40 +7,51 @@ Ferroamp MQTT support sends updates to these topics:
...
 
+## Upgrading from a version before config flow was implemented
+
+1. Remove the integration configuration. i.e this
+   ```yaml
+   sensor:
+     - platform: ferroamp
+   ```
+2. Restart Home Assistant
+3. Remove all Ferroamp entities (they will be re-created once the new setup is completed)
```
Comment (henricm) [inline README.md:42]: Will this make all history for those entities go away?

Comment (argoyle) [reply README.md:42]: I think that as long as the entity IDs remain the same, the history will stay. At least that's what I have seen in my test environment. And I think that most of the sensors have retained their IDs.

**Comment by danielolsson100** [comment]
Hi 
First of all nice work, this is major upgrade of this integration and it's really appreciated that you share the code! 

Note. I tested to migrate to the @argoyle version of the module but it fails on my existing HA running on a rpi4 with the latest version of HA. I therefore created an fresh HA instance running as a Hyper-VM and it works there.

Some reflections in my opinion, I am missing data presentation consistency regarding accuracy.
Example
Temperature: should be presented an an INT. An INT in this case is accurate enough. 
Percentage: should be presented as a FLOAT with one decimal. It is displayed as INT and it's to blunt in this case when you are looking at SoH as an example.
Ampere: should be presented as a FLOAT with one or 2 decimals. It is displayed as INT or FLOAT with to decimals today. 
kwh: should be presented as a FLOAT with one decimal, INT is to blunt of you use the sensor data in a utility meter per day as an example.
I am using these settings above in my fork at https://github.com/danielolsson100/ha-ferroamp

Perpahps we can agree on the data accuracy @henricm ? and migrate my change into your master in a upcoming release ?

I also noted that the grid frequency seems to be missing but that may be by design.
Grid Frequency should be presented with as a FLOAT with 2 decimals and updated directly.
I am using that as a separate sensor from the MQTT Data to allow my house to act as a "good grid citizen" and minimizing the grid load when the grid frequency is below 49.90 :)

**Comment by argoyle** [comment]
> Hi
> First of all nice work, this is major upgrade of this integration and it's really appreciated that you share the code!
> 

Sharing is a no-brainer I would say. 😊 

> Note. I tested to migrate to the @argoyle version of the module but it fails on my existing HA running on a rpi4 with the latest version of HA. I therefore created an fresh HA instance running as a Hyper-VM and it works there.

Did you see any errors in the logs that could explain what went wrong. Would be nice to iron out any upgrade issues before we merge I would say.

> 
> Some reflections in my opinion, I am missing data presentation consistency regarding accuracy.
> Example
> Temperature: should be presented an an INT. An INT in this case is accurate enough.
> Percentage: should be presented as a FLOAT with one decimal. It is displayed as INT and it's to blunt in this case when you are looking at SoH as an example.
> Ampere: should be presented as a FLOAT with one or 2 decimals. It is displayed as INT or FLOAT with to decimals today.
> kwh: should be presented as a FLOAT with one decimal, INT is to blunt of you use the sensor data in a utility meter per day as an example.
> I am using these settings above in my fork at https://github.com/danielolsson100/ha-ferroamp
> 

I have no problem changing this if @henricm agrees.

> Apparent power has a typo in the mdi-icon, should be mdi:transmission-tower
> 

Fixing!

> Perpahps we can agree on the data accuracy @henricm ? and migrate my change into your master in a upcoming release ?
> 
> I also noted that the grid frequency seems to be missing but that may be by design.
> Grid Frequency should be presented with as a FLOAT with 2 decimals and updated directly.
> I am using that as a separate sensor from the MQTT Data to allow my house to act as a "good grid citizen" and minimizing the grid load when the grid frequency is below 49.90 :)

I see no reason to not add the sensor but can you explain more what the reason is for minimizing grid load?

**Comment by henricm** [comment]
Thanks for your comments! Regarding the datatypes - sounds like a good idea to make it more consistent across the sensors. Please open an issue for that and if you have a suggestion please also feel free to submit a PR if you have changes already.
I think it would be good to leave that particular part out of this PR since it already changes a lot. Hope that's ok.

I'm still trying to get my HA devenv up and running so I haven't tried these changes yet. Will see if I can do it later today or the coming week.

**Comment by argoyle** [comment]
I have a few more fixes/changes and full test coverage now but I'll save that for another PR I think. That will add selectors to the services to make it possible to use them in UI-mode in the latest version of Home Assistant as well.

**Comment by argoyle** [comment]
Updated README with information regarding three renamed sensors due to typos being fixed

**Comment by argoyle** [comment]
Any progress @henricm ? 😊

**Comment by Vajomi** [comment]
It would be nice if the development of this project continues :-)

**Comment by henricm** [comment]
Unfortunately no progress from my end I'm afraid. But please feel free to move this forward without me as I cannot make any promises when I'll be able to dive in to this again.

**Comment by argoyle** [comment]
I have been running this code for quite some time now without any problems but I know that @danielolsson100 had some issues running on a Raspberry Pi. It would be good if someone else tried it out as well before merging perhaps.

**Comment by argoyle** [comment]
Maybe you could add me as a contributor to the repository as well @henricm?

**Comment by henricm** [comment]
That's a great idea. Will do that. Thinking we should also merge this PR even though I haven't tested it since these are really good changes.

**Pull request merged**